File: java/src/main/java/br/com/emendes/problems/Permutations.java
package br.com.emendes.problems;

import java.util.ArrayList;
import java.util.List;

/**
 * Dado um array nums de números inteiros distintos, retorne todas as permutações possíveis.<br>
 * Você pode retornar a resposta em qualquer ordem.<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= nums.length <= 6<br>
 * -10 <= nums[i] <= 10<br>
 * Todos os inteiros em nums são únicos.
 */
public class Permutations {

  public List<List<Integer>> permute(int[] nums) {
    List<List<Integer>> ans = new ArrayList<>();
    generatePermutations(0, nums, ans);
    return ans;
  }

  private void generatePermutations(int ind, int[] nums, List<List<Integer>> ans) {
    if (ind == nums.length) {
      List<Integer> permutation = new ArrayList<>();
      for (int num : nums) {
        permutation.add(num);
      }
      ans.add(permutation);
      return;
    }

    for (int i = ind; i < nums.length; i++) {
      swap(i, ind, nums);
      generatePermutations(ind + 1, nums, ans);
      swap(i, ind, nums);
    }
  }

  private void swap(int i, int j, int[] nums) {
    int t = nums[i];
    nums[i] = nums[j];
    nums[j] = t;
  }

}

File: javascript/problems/Permutations.js
/**
 * Dado um array nums de inteiros distintos, retorne todas as permutações possíveis.
 * Voçê pode retornar a solução em qualquer ordem.
 *
 * Restrições:
 *
 * 1 <= nums.length <= 6
 *
 * -10 <= nums[i] <= 10
 *
 * Todos os números de nums são distintos
 *
 * @param {number[]} nums
 * @return {number[][]}
 */
const permute = function (nums) {
  const permutations = [];
  generatePermutations(0, nums, permutations);

  return permutations;
};

const generatePermutations = (index, permutation, permutations) => {
  if (index === permutation.length) {
    permutations.push([...permutation]);
    return;
  }
  for (let i = index; i < permutation.length; i++) {
    swap(permutation, index, i);
    generatePermutations(index + 1, permutation, permutations);
    swap(permutation, i, index);
  }
};

const swap = (arr, i1, i2) => {
  let temp = arr[i1];
  arr[i1] = arr[i2];
  arr[i2] = temp;
};

/**
 * First solution.
 */
const firstSolution = function (nums) {
  const permutations = [];
  generatePermutations2(nums, [], permutations);

  return permutations;
};

const generatePermutations2 = (nums, permutation, permutations) => {
  console.log(permutation);

  if (permutation.length === nums.length) {
    permutations.push([...permutation]);
    return;
  }
  for (let index = 0; index < nums.length; index++) {
    const element = nums[index];
    if (element !== ".") {
      permutation.push(element);
      nums[index] = ".";
      generatePermutations2(nums, permutation, permutations);
      nums[index] = element;
      permutation.pop();
    }
  }
};
